File: Persimmon.VisualStudio.TestExplorer2/VSTestAdapter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestPlatform.ObjectModel;
using Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter;
using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;

namespace Persimmon.VisualStudio.TestExplorer2
{

    [FileExtension(".dll")]
    [ExtensionUri("executor://persimmon.visualstudio.testexplorer")]
    [DefaultExecutorUri("executor://persimmon.visualstudio.testexplorer")]
    public class VSTestAdapter : ITestDiscoverer, ITestExecutor
    {
        private VSTestRunner runner = new VSTestRunner();

        public void Cancel()
        {
        }

        public void DiscoverTests(IEnumerable<string> sources, IDiscoveryContext discoveryContext, IMessageLogger logger, ITestCaseDiscoverySink discoverySink)
        {
            try
            {
                logger.SendMessage(TestMessageLevel.Informational, sources.ToString());
                runner.DiscoverTests(sources.ToList(), discoverySink);
            }
            catch (Exception e)
            {
                logger.SendMessage(TestMessageLevel.Error, e.ToString());
            }
        }

        public void RunTests(IEnumerable<string> sources, IRunContext runContext, IFrameworkHandle frameworkHandle)
        {
            runner.RunTests(sources.Distinct().ToList(), frameworkHandle);
        }

        public void RunTests(IEnumerable<Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase> tests, IRunContext runContext, IFrameworkHandle frameworkHandle)
        {
            this.RunTests(tests.Select(c => c.Source), runContext, frameworkHandle);
        }
    }
}

File: Persimmon.VisualStudio.TestExplorer/VSTestAdapter.fs
namespace Persimmon.VisualStudio.TestExplorer

open Microsoft.VisualStudio.TestPlatform.ObjectModel
open Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter
open Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging
open System
open System.Reflection

[<FileExtension(".dll")>]
[<ExtensionUri(Constant.extensionUri)>]
[<DefaultExecutorUri(Constant.extensionUri)>]
type VSTestAdapter() =

  let version = Assembly.GetExecutingAssembly().GetName().Version.ToString()

  let runner = VSTestRunner()

  interface ITestDiscoverer with
    member this.DiscoverTests(sources: string seq, context: IDiscoveryContext,logger: IMessageLogger, sink: ITestCaseDiscoverySink) =
      try
        logger.SendMessage(TestMessageLevel.Informational, sprintf "Persimmon Test Explorer %s discovering tests is started" version)
        try
          runner.DiscoverTests(sources, sink)
        with e ->
          logger.SendMessage(TestMessageLevel.Error, e.ToString())
      finally
        logger.SendMessage(TestMessageLevel.Informational, sprintf "Persimmon Test Explorer %s discovering tests is finished" version)
  interface ITestExecutor with
    member this.RunTests(tests: TestCase seq, runContext: IRunContext, handle: IFrameworkHandle) =
      (this :> ITestExecutor).RunTests(tests |> Seq.map (fun c -> c.Source), runContext, handle)
    member this.RunTests(sources: string seq, _: IRunContext, handle: IFrameworkHandle) =
      runner.RunTests(sources, handle)
    member __.Cancel() =
      ()
